

关闭流的正确方法
try{
   input = new FileInputStream("c:\\data\\input-text.txt");
   doSomethingWithData(data);
} finally {
  try{
    if(input != null) input.close();   //关闭流时可能发生异常，需要再次捕获
  } catch(IOException e){
    //do something, or ignore.
  }
}

正确处理 InputStream（或 OutputStream）迭代的代码，如您所见，可以打赌非常丑陋，一旦您在那里也获得了正确的异常处理。
这种丑陋的异常处理代码在您的代码中传播并一遍又一遍地重复并不是特别好。如果有人急着走捷径，跳过异常处理怎么办？

此外，假设首先从 doSomethingWithData() 抛出异常。第一个 catch 子句将捕获该异常，然后 finally 子句将尝试关闭 InputStream。
但是，如果 input.close() 方法也抛出异常会发生什么？这两个异常中的哪一个应该向上传播到调用堆栈？

幸运的是，有一种方法可以解决这个问题。该解决方案称为“Exception Handling Templates”。创建一个异常处理模板，在使用后正确关闭您的流。此模板编写一次，
并在整个代码中重复使用。好看又简单。要了解更多信息，请访问 Java 中的异常处理模板。


Java IO Exception Handling From Java 7
从 Java 7 开始，Java 包含一个新的异常处理机制，称为“try with resources”。这种异常处理机制特别针对在使用使用后需要正确关闭的资源时的异常处理，
如 InputStream、OutputStream 等。 你可以在我的关于在 Java 7 中尝试资源的教程中阅读更多关于它的信息。
```
try(FileInputStream input = new FileInputStream("file.txt")) {
   doSomethingWithData(data);
    }
```